### PR TITLE
feat(data): pass request options to custom header functions

### PR DESCRIPTION
**Pull request by david-mcafee**
#### Description of changes

This PR updates the API GraphQL category, as well as the AWSAppSyncRealTimeProvider, to send request options to custom header functions, which will allow customers that use custom header functions to access request information to sign their requests cryptographically. Request options are available when 1) generating the API client, 2) executing individual model operations, and 3) when creating subscriptions.

#### Description of how you validated changes
1. Manual testing
2. Added unit tests

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/api-graphql/__tests__/generateClient.test.ts [hunk 1]
```diff
@@ -66,23 +66,23 @@ function makeAppSyncStreams() {
  * replaces fields that are likely to change between calls (or library version revs)
  * with static values. When possible, on the unpredicable portions of these values
  * are replaced.
- *
```
Comment (david-mcafee) [inline packages/api-graphql/__tests__/generateClient.test.ts:69]: Looks like this file was previously updated without Prettier - I've added comments below to easily identify the added tests.

File: packages/api-graphql/__tests__/generateClient.test.ts [hunk 2]
```diff
@@ -2795,7 +2795,46 @@ describe('generateClient', () => {
 				description: 'something something',
 			});
 
-			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot()
+			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();
```
Comment (david-mcafee) [inline packages/api-graphql/__tests__/generateClient.test.ts:2798]: Added tests here

File: packages/api-graphql/__tests__/generateClient.test.ts [hunk 3]
```diff
@@ -2894,7 +2933,53 @@ describe('generateClient', () => {
 				}
 			);
 
-			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot()
+			expect(normalizePostGraphqlCalls(spy)).toMatchSnapshot();
```
Comment (david-mcafee) [inline packages/api-graphql/__tests__/generateClient.test.ts:2936]: Added tests here

File: packages/api-graphql/__tests__/generateClient.test.ts [hunk 4]
```diff
@@ -3313,6 +3398,50 @@ describe('generateClient', () => {
 			});
 		});
 
+		test('can subscribe to onCreate() - with a custom header function that accepts requestOptions', done => {
```
Comment (david-mcafee) [inline packages/api-graphql/__tests__/generateClient.test.ts:3401]: Added tests here

File: packages/api-graphql/__tests__/AWSAppSyncRealTimeProvider.test.ts [hunk 5]
```diff
@@ -1104,6 +1104,27 @@ describe('AWSAppSyncRealTimeProvider', () => {
...
+					test('authenticating with AWS_LAMBDA/custom w/ custom header function that accepts request options', async () => {
+						expect.assertions(1);
+
+						provider
+							.subscribe({
+								appSyncGraphqlEndpoint: 'ws://localhost:8080',
+								authenticationType: 'none',
+								additionalHeaders: async requestOptions => ({
+									Authorization: 'test',
+								}),
```
Comment (svidgen) [inline packages/api-graphql/__tests__/AWSAppSyncRealTimeProvider.test.ts:1116]: Can we assert that this actually gets called?

Comment (david-mcafee) [reply packages/api-graphql/__tests__/AWSAppSyncRealTimeProvider.test.ts]: Done

**Comment by stocaaro** [comment]
The change looks fine. For posterity's sake, can you add a description to this PR that captures what customer behavior is changing with this diff and confirm that you've tested the behavior in a sample app? Happy to approve this once the description is filled in a bit.

**Comment by david-mcafee** [comment]
> The change looks fine. For posterity's sake, can you add a description to this PR that captures what customer behavior is changing with this diff and confirm that you've tested the behavior in a sample app? Happy to approve this once the description is filled in a bit.

Good callout! This has been added.

**Review by stocaaro — APPROVED** [review]

**Review by svidgen — APPROVED** [review]

**Pull request merged**